crates/uv-auth/src/cache.rs

```diff
@@ -1,43 +1,16 @@
 use std::sync::Arc;
 use std::{collections::HashMap, sync::Mutex};
 
-use crate::credentials::Credentials;
+use crate::credentials::{Credentials, Username};
 use crate::NetLoc;
 
 use tracing::trace;
 use url::Url;
 
-type CacheKey = (NetLoc, Option<String>);
-
 pub struct CredentialsCache {
-    store: Mutex<HashMap<CacheKey, Arc<Credentials>>>,
-}
-
-#[derive(Debug, Clone)]
-pub enum CheckResponse {
-    /// The given credentials should be used and are not present in the cache.
-    Uncached(Arc<Credentials>),
-    /// Credentials were found in the cache.
-    Cached(Arc<Credentials>),
-    // Credentials were not found in the cache and none were provided.
-    None,
-}
-
-impl CheckResponse {
-    /// Retrieve the credentials, if any.
-    pub fn get(&self) -> Option<&Credentials> {
-        match self {
-            Self::Cached(credentials) => Some(credentials.as_ref()),
-            Self::Uncached(credentials) => Some(credentials.as_ref()),
-            Self::None => None,
-        }
-    }
-
-    /// Returns true if there are credentials with a password.
-    pub fn is_authenticated(&self) -> bool {
-        self.get()
-            .is_some_and(|credentials| credentials.password().is_some())
-    }
+    realms: Mutex<HashMap<(NetLoc, Username), Arc<Credentials>>>,
+    #[allow(clippy::type_complexity)]
+    urls: Mutex<Vec<((Url, Username), Arc<Credentials>)>>,
```

Comment:
This doesn't seem ideal, we need to scan all of the authenticated URLs on each request? I presume we could make this a `BTreeSet` and use `Range` to help search?

Reply:
We only insert new URLs into the cache though if we fail to find authentication in the cache already though, so hopefully the size is relatively small?

Reply:
I think you want a trie?

Reply:
Yeah maybe. I'm questioning complexity vs performance here.